File: frontend/src/api/jobs.ts
import client from './client'

// --- Types ---

export type JobType = 'eval' | 'index_build'
export type JobStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'

export interface JobResponse {
  id: number
  job_type: JobType
  run_id: number
  status: JobStatus
  progress: number
  config_json: Record<string, unknown>
  result_json: Record<string, unknown>
  error_message: string | null
  started_at: string | null
  ended_at: string | null
  created_at: string
}

export interface EvalJobRequest {
  run_id: number
  checkpoint?: string
  bit_lengths?: number[]
  k_values?: number[]
}

export interface IndexBuildJobRequest {
  run_id: number
  checkpoint?: string
  image_dir?: string | null
  captions_file?: string | null
  thumbnail_size?: number
  batch_size?: number
}

export interface SearchResult {
  rank: number
  index: number
  score: number
  thumbnail_b64: string | null
  caption: string | null
}

export interface SearchResponse {
  results: SearchResult[]
  query_hash: number[]
  search_time_ms: number
  method: string
  bit_length: number
  query?: string
}

// --- Job API ---

export const createEvalJob = async (data: EvalJobRequest): Promise<JobResponse> => {
  const response = await client.post('/jobs/eval', data)
  return response.data
}

export const createIndexBuildJob = async (data: IndexBuildJobRequest): Promise<JobResponse> => {
  const response = await client.post('/jobs/index-build', data)
  return response.data
}

export const getJob = async (jobId: number): Promise<JobResponse> => {
  const response = await client.get(`/jobs/${jobId}`)
  return response.data
}

export const listJobs = async (params?: {
  job_type?: JobType
  run_id?: number
}): Promise<JobResponse[]> => {
  const response = await client.get('/jobs', { params })
  return response.data
}

export const cancelJob = async (jobId: number): Promise<{ status: string }> => {
  const response = await client.post(`/jobs/${jobId}/cancel`)
  return response.data
}

// --- Search API ---

export const searchByText = async (params: {
  query: string
  index_path: string
  checkpoint_path: string
  bit_length?: number
  top_k?: number
  method?: string
}): Promise<SearchResponse> => {
  const formData = new FormData()
  formData.append('query', params.query)
  formData.append('index_path', params.index_path)
  formData.append('checkpoint_path', params.checkpoint_path)
  formData.append('bit_length', String(params.bit_length ?? 64))
  formData.append('top_k', String(params.top_k ?? 20))
  formData.append('method', params.method ?? 'hamming')

  const response = await client.post('/search/text', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return response.data
}

export const searchByImage = async (params: {
  image: File
  index_path: string
  checkpoint_path: string
  bit_length?: number
  top_k?: number
  method?: string
}): Promise<SearchResponse> => {
  const formData = new FormData()
  formData.append('image', params.image)
  formData.append('index_path', params.index_path)
  formData.append('checkpoint_path', params.checkpoint_path)
  formData.append('bit_length', String(params.bit_length ?? 64))
  formData.append('top_k', String(params.top_k ?? 20))
  formData.append('method', params.method ?? 'hamming')

  const response = await client.post('/search/image', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return response.data
}

File: backend/api/jobs.py
"""REST API endpoints for background job management."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.models.database import get_session
from backend.schemas.job import (
    EvalJobRequest,
    IndexBuildJobRequest,
    JobProgressUpdate,
    JobResponse,
)
from backend.services.job_manager import job_manager
from shared.schemas import JobType

router = APIRouter(prefix="/api/jobs", tags=["jobs"])


@router.post("/eval", response_model=JobResponse, status_code=201)
async def create_eval_job(
    body: EvalJobRequest,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> JobResponse:
    """Start an evaluation job for a run."""
    try:
        job = await job_manager.create_eval_job(
            run_id=body.run_id,
            config={
                "checkpoint": body.checkpoint,
                "bit_lengths": body.bit_lengths,
                "k_values": body.k_values,
            },
            session=session,
        )
        return JobResponse.model_validate(job)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.post("/index-build", response_model=JobResponse, status_code=201)
async def create_index_build_job(
    body: IndexBuildJobRequest,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> JobResponse:
    """Start an index build job for a run."""
    try:
        job = await job_manager.create_index_build_job(
            run_id=body.run_id,
            config={
                "checkpoint": body.checkpoint,
                "image_dir": body.image_dir,
                "captions_file": body.captions_file,
                "thumbnail_size": body.thumbnail_size,
                "batch_size": body.batch_size,
            },
            session=session,
        )
        return JobResponse.model_validate(job)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/{job_id}", response_model=JobResponse)
async def get_job(
    job_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> JobResponse:
    """Get job details and current progress."""
    job = await job_manager.get_job(job_id, session)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return JobResponse.model_validate(job)


@router.get("", response_model=list[JobResponse])
async def list_jobs(
    session: Annotated[AsyncSession, Depends(get_session)],
    job_type: JobType | None = Query(default=None),
    run_id: int | None = Query(default=None),
) -> list[JobResponse]:
    """List jobs with optional filters."""
    jobs = await job_manager.list_jobs(session=session, job_type=job_type, run_id=run_id)
    return [JobResponse.model_validate(j) for j in jobs]


@router.post("/{job_id}/progress", response_model=JobResponse)
async def update_job_progress(
    job_id: int,
    body: JobProgressUpdate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> JobResponse:
    """Internal endpoint: update job progress from subprocess."""
    job = await job_manager.update_progress(
        job_id=job_id,
        progress=body.progress,
        session=session,
        status=body.status,
        result_json=body.result_json,
        error_message=body.error_message,
    )
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return JobResponse.model_validate(job)


@router.post("/{job_id}/cancel")
async def cancel_job(
    job_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> dict[str, str]:
    """Cancel a running job."""
    cancelled = await job_manager.cancel_job(job_id, session)
    if not cancelled:
        raise HTTPException(status_code=404, detail="Job not found or not running")
    return {"status": "cancelled"}
